File: packages/app/android/src/test/java/io/invertase/firebase/common/RNFBHandleMapTest.java
package io.invertase.firebase.common;

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.Before;
import org.junit.Test;

/**
 * JVM coverage for {@link RNFBHandleMap}. Tests must not invoke methods on stored values — the map
 * only moves pointers.
 */
public class RNFBHandleMapTest {

  private RNFBHandleMap<Integer, Object> map;

  @Before
  public void setUp() {
    map = new RNFBHandleMap<>();
  }

  @Test
  public void putGetTake_happyPath() throws Exception {
    Object handle = new Object();
    map.put(1, handle);

    assertSame(handle, map.get(1));
    assertSame(handle, map.take(1));
    assertNull(map.get(1));
  }

  @Test
  public void take_missingKey_returnsNull() {
    assertNull(map.take(99));
    assertNull(map.get(99));
  }

  @Test
  public void takeAll_returnsSnapshotAndLeavesMapEmpty() throws Exception {
    Object a = new Object();
    Object b = new Object();
    map.put(1, a);
    map.put(2, b);

    List<Object> snapshot = map.takeAll();

    assertEquals(2, snapshot.size());
    assertTrue(snapshot.contains(a));
    assertTrue(snapshot.contains(b));
    assertNull(map.get(1));
    assertNull(map.get(2));
    assertTrue(map.takeAll().isEmpty());
  }

  @Test
  public void put_occupiedId_throwsCollision() throws Exception {
    Object first = new Object();
    map.put(1, first);
    try {
      map.put(1, new Object());
      fail("expected RNFBHandleCollisionException");
    } catch (RNFBHandleCollisionException e) {
      assertTrue(e.getMessage().contains("1"));
      assertSame(first, map.get(1));
    }
  }

  @Test
  public void putIfAbsent_whenFree_storesAndReturnsTrue() throws Exception {
    Object handle = new Object();
    assertTrue(map.putIfAbsent(1, handle));
    assertSame(handle, map.get(1));
  }

  @Test
  public void putIfAbsent_whenOccupied_keepsExistingAndReturnsFalse() throws Exception {
    Object first = new Object();
    Object second = new Object();
    map.put(1, first);
    assertFalse(map.putIfAbsent(1, second));
    assertSame(first, map.get(1));
  }

  @Test
  public void putIfAbsentOrSame_whenAbsent_storesAndReturnsTrue() {
    Object handle = new Object();
    assertTrue(map.putIfAbsentOrSame(1, handle));
    assertSame(handle, map.get(1));
  }

  @Test
  public void putIfAbsentOrSame_whenSameInstance_returnsTrue() throws Exception {
    Object handle = new Object();
    map.put(1, handle);
    assertTrue(map.putIfAbsentOrSame(1, handle));
    assertSame(handle, map.get(1));
  }

  @Test
  public void putIfAbsentOrSame_whenDifferent_returnsFalse() throws Exception {
    Object first = new Object();
    Object second = new Object();
    map.put(1, first);
    assertFalse(map.putIfAbsentOrSame(1, second));
    assertSame(first, map.get(1));
  }

  @Test
  public void putReplacing_whenFree_storesAndReturnsNull() throws Exception {
    Object handle = new Object();
    assertNull(map.putReplacing(1, handle));
    assertSame(handle, map.get(1));
  }

  @Test
  public void putReplacing_whenOccupied_replacesLastWins() throws Exception {
    Object first = new Object();
    Object second = new Object();
    map.put(1, first);
    assertSame(first, map.putReplacing(1, second));
    assertSame(second, map.get(1));
  }

  @Test
  public void put_afterTake_allowsReuse() throws Exception {
    Object first = new Object();
    Object second = new Object();
    map.put(1, first);
    assertSame(first, map.take(1));
    map.put(1, second);
    assertSame(second, map.get(1));
  }

  @Test
  public void take_sameIdFromTwoThreads_onlyOneReturnsNonNull() throws Exception {
    Object handle = new Object();
    map.put(1, handle);

    CountDownLatch start = new CountDownLatch(1);
    CountDownLatch done = new CountDownLatch(2);
    AtomicReference<Object> first = new AtomicReference<>();
    AtomicReference<Object> second = new AtomicReference<>();

    Thread t1 =
        new Thread(
            () -> {
              try {
                start.await();
                first.set(map.take(1));
              } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
              } finally {
                done.countDown();
              }
            });
    Thread t2 =
        new Thread(
            () -> {
              try {
                start.await();
                second.set(map.take(1));
              } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
              } finally {
                done.countDown();
              }
            });

    t1.start();
    t2.start();
    start.countDown();
    assertTrue(done.await(5, TimeUnit.SECONDS));

    Object a = first.get();
    Object b = second.get();
    boolean onlyFirst = a == handle && b == null;
    boolean onlySecond = b == handle && a == null;
    assertTrue(onlyFirst || onlySecond);
    assertNull(map.get(1));
  }

  @Test
  public void putReplacing_sameIdFromTwoThreads_lastWins() throws Exception {
    Object first = new Object();
    Object second = new Object();
    Object third = new Object();
    map.put(1, first);

    CountDownLatch start = new CountDownLatch(1);
    CountDownLatch done = new CountDownLatch(2);
    AtomicReference<Object> prevA = new AtomicReference<>();
    AtomicReference<Object> prevB = new AtomicReference<>();

    Thread t1 =
        new Thread(
            () -> {
              try {
                start.await();
                prevA.set(map.putReplacing(1, second));
              } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
              } finally {
                done.countDown();
              }
            });
    Thread t2 =
        new Thread(
            () -> {
              try {
                start.await();
                prevB.set(map.putReplacing(1, third));
              } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
              } finally {
                done.countDown();
              }
            });

    t1.start();
    t2.start();
    start.countDown();
    assertTrue(done.await(5, TimeUnit.SECONDS));

    Object winner = map.get(1);
    assertTrue(winner == second || winner == third);
    Object returnedFirst = prevA.get() == first ? prevA.get() : prevB.get();
    assertSame(first, returnedFirst);
  }
}

File: packages/app/android/src/test/java/io/invertase/firebase/common/RNFBHandleMapTest.kt
package io.invertase.firebase.common

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertSame
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference

/**
 * JVM coverage for [RNFBHandleMap]. Tests must not invoke methods on stored values — the map only
 * moves pointers.
 */
class RNFBHandleMapTest {
  private lateinit var map: RNFBHandleMap<Int, Any>

  @Before
  fun setUp() {
    map = RNFBHandleMap()
  }

  @Test
  fun putGetTake_happyPath() {
    val handle = Any()
    map.put(1, handle)

    assertSame(handle, map.get(1))
    assertSame(handle, map.take(1))
    assertNull(map.get(1))
  }

  @Test
  fun take_missingKey_returnsNull() {
    assertNull(map.take(99))
    assertNull(map.get(99))
  }

  @Test
  fun takeAll_returnsSnapshotAndLeavesMapEmpty() {
    val a = Any()
    val b = Any()
    map.put(1, a)
    map.put(2, b)

    val snapshot = map.takeAll()

    assertEquals(2, snapshot.size)
    assertTrue(snapshot.contains(a))
    assertTrue(snapshot.contains(b))
    assertNull(map.get(1))
    assertNull(map.get(2))
    assertTrue(map.takeAll().isEmpty())
  }

  @Test
  fun put_occupiedId_throwsCollision() {
    val first = Any()
    map.put(1, first)
    try {
      map.put(1, Any())
      fail("expected RNFBHandleCollisionException")
    } catch (exception: RNFBHandleCollisionException) {
      assertTrue(exception.message!!.contains("1"))
      assertSame(first, map.get(1))
    }
  }

  @Test
  fun put_nullValue_stillTreatsIdAsOccupied() {
    val nullableMap = RNFBHandleMap<Int, Any?>()
    nullableMap.put(1, null)

    try {
      nullableMap.put(1, Any())
      fail("expected RNFBHandleCollisionException")
    } catch (exception: RNFBHandleCollisionException) {
      assertEquals("Handle id already registered: 1", exception.message)
      assertNull(nullableMap.get(1))
    }
  }

  @Test
  fun putIfAbsent_whenFree_storesAndReturnsTrue() {
    val handle = Any()
    assertTrue(map.putIfAbsent(1, handle))
    assertSame(handle, map.get(1))
  }

  @Test
  fun putIfAbsent_whenOccupied_keepsExistingAndReturnsFalse() {
    val first = Any()
    val second = Any()
    map.put(1, first)
    assertFalse(map.putIfAbsent(1, second))
    assertSame(first, map.get(1))
  }

  @Test
  fun putIfAbsent_nullableExistingValue_isOccupied() {
    val nullableMap = RNFBHandleMap<Int, Any?>()
    nullableMap.put(1, null)

    assertFalse(nullableMap.putIfAbsent(1, Any()))
    assertNull(nullableMap.get(1))
  }

  @Test
  fun putIfAbsentOrSame_whenAbsent_storesAndReturnsTrue() {
    val handle = Any()
    assertTrue(map.putIfAbsentOrSame(1, handle))
    assertSame(handle, map.get(1))
  }

  @Test
  fun putIfAbsentOrSame_whenSameInstance_returnsTrue() {
    val handle = Any()
    map.put(1, handle)
    assertTrue(map.putIfAbsentOrSame(1, handle))
    assertSame(handle, map.get(1))
  }

  @Test
  fun putIfAbsentOrSame_whenDifferent_returnsFalse() {
    val first = Any()
    val second = Any()
    map.put(1, first)
    assertFalse(map.putIfAbsentOrSame(1, second))
    assertSame(first, map.get(1))
  }

  @Test
  fun putIfAbsentOrSame_sameNullInstance_returnsTrue() {
    val nullableMap = RNFBHandleMap<Int, Any?>()
    nullableMap.put(1, null)

    assertTrue(nullableMap.putIfAbsentOrSame(1, null))
    assertNull(nullableMap.get(1))
  }

  @Test
  fun putIfAbsentOrSame_nullExistingAndDifferentHandle_returnsFalse() {
    val nullableMap = RNFBHandleMap<Int, Any?>()
    nullableMap.put(1, null)

    assertFalse(nullableMap.putIfAbsentOrSame(1, Any()))
    assertNull(nullableMap.get(1))
  }

  @Test
  fun putReplacing_whenFree_storesAndReturnsNull() {
    val handle = Any()
    assertNull(map.putReplacing(1, handle))
    assertSame(handle, map.get(1))
  }

  @Test
  fun putReplacing_whenOccupied_replacesLastWins() {
    val first = Any()
    val second = Any()
    map.put(1, first)
    assertSame(first, map.putReplacing(1, second))
    assertSame(second, map.get(1))
  }

  @Test
  fun put_afterTake_allowsReuse() {
    val first = Any()
    val second = Any()
    map.put(1, first)
    assertSame(first, map.take(1))
    map.put(1, second)
    assertSame(second, map.get(1))
  }

  @Test
  fun takeIf_whenPredicateTrue_removesAndReturns() {
    val handle = Any()
    map.put(1, handle)
    assertSame(handle, map.takeIf(1) { it === handle })
    assertNull(map.get(1))
  }

  @Test
  fun takeIf_whenPredicateFalse_leavesMapping() {
    val handle = Any()
    map.put(1, handle)
    assertNull(map.takeIf(1) { false })
    assertSame(handle, map.get(1))
  }

  @Test
  fun takeIf_whenMissing_returnsNull() {
    assertNull(map.takeIf(99) { true })
  }

  @Test
  fun take_sameIdFromTwoThreads_onlyOneReturnsNonNull() {
    val handle = Any()
    map.put(1, handle)

    val start = CountDownLatch(1)
    val done = CountDownLatch(2)
    val first = AtomicReference<Any>()
    val second = AtomicReference<Any>()

    val firstThread =
      Thread {
        try {
          start.await()
          first.set(map.take(1))
        } catch (_: InterruptedException) {
          Thread.currentThread().interrupt()
        } finally {
          done.countDown()
        }
      }
    val secondThread =
      Thread {
        try {
          start.await()
          second.set(map.take(1))
        } catch (_: InterruptedException) {
          Thread.currentThread().interrupt()
        } finally {
          done.countDown()
        }
      }

    firstThread.start()
    secondThread.start()
    start.countDown()
    assertTrue(done.await(5, TimeUnit.SECONDS))

    val a = first.get()
    val b = second.get()
    val onlyFirst = a === handle && b == null
    val onlySecond = b === handle && a == null
    assertTrue(onlyFirst || onlySecond)
    assertNull(map.get(1))
  }

  @Test
  fun putReplacing_sameIdFromTwoThreads_lastWins() {
    val first = Any()
    val second = Any()
    val third = Any()
    map.put(1, first)

    val start = CountDownLatch(1)
    val done = CountDownLatch(2)
    val previousA = AtomicReference<Any>()
    val previousB = AtomicReference<Any>()

    val firstThread =
      Thread {
        try {
          start.await()
          previousA.set(map.putReplacing(1, second))
        } catch (_: InterruptedException) {
          Thread.currentThread().interrupt()
        } finally {
          done.countDown()
        }
      }
    val secondThread =
      Thread {
        try {
          start.await()
          previousB.set(map.putReplacing(1, third))
        } catch (_: InterruptedException) {
          Thread.currentThread().interrupt()
        } finally {
          done.countDown()
        }
      }

    firstThread.start()
    secondThread.start()
    start.countDown()
    assertTrue(done.await(5, TimeUnit.SECONDS))

    val winner = map.get(1)
    assertTrue(winner === second || winner === third)
    val returnedFirst = if (previousA.get() === first) previousA.get() else previousB.get()
    assertSame(first, returnedFirst)
  }
}
